src/main/java/com/coveo/spillway/storage/sliding/AsyncSlidingLimitUsageStorage.java

```diff
@@ -0,0 +1,108 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.coveo.spillway.storage.sliding;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Timer;
+
+import com.coveo.spillway.limit.LimitKey;
+import com.coveo.spillway.storage.LimitUsageStorage;
+import com.coveo.spillway.storage.utils.AddAndGetRequest;
+import com.coveo.spillway.storage.utils.SlidingCacheSynchronisation;
+
+/**
+ * An asynchronous sliding window implementation of {@link LimitUsageStorage}.
+ * <p>
+ * This storage internally uses a {@link InMemorySlidingStorage} as cache and performs
+ * asynchronous calls to the distributed storage to share information.
+ * <p>
+ * This it particularly useful when using a database over the network as
+ * the queries are not slowed down by any external problems.
+ * <p>
+ * This storage does not sync with the main server each time a query arrives. A configurable
+ * timeout launches the synchronization.
+ * <p>
+ * The advantage of this method is that the load on the network and on the external
+ * storage is considerably reduced at the cost of a less precise throttling. We
+ * recommend to set a relatively small time between each synchronization to avoid
+ * big differences between the throttling instances.
+ *
+ * @author Emile Fugulin
+ * @since 1.1.0
+ */
+public class AsyncSlidingLimitUsageStorage implements LimitUsageStorage {
+  private final LimitUsageStorage wrappedLimitUsageStorage;
+  private InMemorySlidingStorage cache;
+  private Timer timer;
+
+  public AsyncSlidingLimitUsageStorage(
+      LimitUsageStorage wrappedLimitUsageStorage,
+      Duration timeBetweenSynchronisations,
+      Duration cacheRetentionTime,
+      Duration cacheSlideSize) {
+    this(
+        wrappedLimitUsageStorage,
+        timeBetweenSynchronisations,
+        Duration.ofMillis(0),
+        cacheRetentionTime,
+        cacheSlideSize);
+  }
+
+  /*package*/ AsyncSlidingLimitUsageStorage(
+      LimitUsageStorage wrappedLimitUsageStorage,
+      Duration timeBetweenSynchronisations,
+      Duration delayBeforeFirstSync,
+      Duration cacheRetentionTime,
+      Duration cacheSlideSize) {
+    this.wrappedLimitUsageStorage = wrappedLimitUsageStorage;
+    this.cache = new InMemorySlidingStorage(cacheRetentionTime, cacheSlideSize);
```

Comment:
I found that it is easier not to mock the InMemory storage and do more FT style tests.